[Core] Add a `store_names` property to the bot, thats checked before storing username and nicknames as well as a command to delete currently stored names

Drapersniper opened the pull request:
Stop storing names by default and add toggle to store names and command to delete stored names

### Type

- [ ] Bugfix
- [x] Enhancement
- [x] New feature

### Description of the changes

Drapersniper committed:
Stop storing names by default and add toggle to store names and command to delete stored names

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
black, and put this in the right place

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Dav-Git commented:
Why though?

Drapersniper commented:
1 - Names and User names are User data, so to comply with discord you need to take some extra steps in handling these
2 - Some users may not wish to store names and usernames specially larger bots
3 - Some bot owners may not even be away that these are stored.

Drapersniper committed:
...

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
...

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

SharkyTheKing commented:
I would love to see this, not everyone wants to store nickname/username information on the bot's members. For large bots/servers storing names adds extra load for the person/VPS hosting the bot, being able to toggle it off/on as a core functionality would be beneficial. Allowing the bot owner to decide if they're fine with storing names or not is great, especially since they'll be able to judge if they want that type of functionality or if they can handle that functionality alongside whatever else they're using.

Drapersniper commented:
Okay ... so ... upon further inspection i hit some issues when doing the follow up PR so will be changing the implementation slighly on this one

Drapersniper committed:
...

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
Address modlog and address bank

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
welp

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
welp 2.0

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
welp v3

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
welp v4 is here

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper commented:
Okay because life is hard.....

I've added `Red.store_names` attribute to the bot,
This attribute is now checked before storing names in Bank, Modlog and Mod.

I've added the following commands:
`[p]bankset deletenames` to delete the usernames stored in Bank config
`[p]modset deletenames` to delete the usernames stored in Mod config
`[p]modlogset deletenames` to delete the usernames stored in ModLog config

`[p]set storenames` to toggle the state of this config that by default is off
`[p]set deletenames` to delete which delete the stored names and nicknames for all the 3 cogs we currently store names for.


since `bot.store_names` is now a property in the bot, CC can also use it going forward to know whether the bot owner wishes to store username/nicknames.

Drapersniper committed:
forgot tests existed

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Drapersniper committed:
nothing to see here

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Bakersbakebread commented:
This PR leaves much to be desired, but is definitely a step in the right direction. 

A user ID is not considered End User Data but a Name is, therefore I would like to see an implementation with use of this PR from bot devs/CC's where there is express consent given from the End User as per TOS. 

Having this opt-in is a **must** as the accountability falls on the implementer. We should not be storing EUD beyond the scope of the application, again per TOS. Names seems redundant in my opinion as this information is available by the API - any information returned by the API is also the sole responsibility of Discord and not bot devs/CC's respectively.

Drapersniper committed:
Merge branch 'V3/develop' into mod-names

Drapersniper committed:
Merge branch 'V3/develop' into mod-names

Drapersniper committed:
update PR to new config naming

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Bakersbakebread commented on redbot/cogs/mod/settings.py:
```diff
@@ -221,3 +222,21 @@ async def defaultdays(self, ctx: commands.Context, days: int = 0):
+    async def deletenames(self, ctx: commands.Context) -> None:
+        """Delete all stored names and usernames"""
+
+        async with self.settings._get_base_group(self.settings.MEMBER).all() as mod_member_data:
```
```
Traceback (most recent call last):
  File "a:\projects\venvs\testvenv\lib\site-packages\discord\ext\commands\core.py", line 83, in wrapped
    ret = await coro(*args, **kwargs)
  File "c:\users\seant\desktop\ruinedred\redbot\cogs\mod\settings.py", line 231, in deletenames
    async with self.settings._get_base_group(self.settings.MEMBER).all() as mod_member_data:
AttributeError: 'Mod' object has no attribute 'settings'
```

Drapersniper committed:
update PR to new config naming aaaaaaaaaaaaaaaaaaaaaaaaaaaaa

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Bakersbakebread commented on redbot/core/core_commands.py:
```diff
@@ -976,6 +979,70 @@ async def setdescription(self, ctx: commands.Context, *, description: str = ""):
+        Default is Off
+        """
+        ctx.bot._store_names = not ctx.bot._store_names
+        await ctx.bot._config.store_names.set(ctx.bot.store_names)
```
A DateTime object is being passed here - can't trace where from? I had zero config setup

Drapersniper committed:
aaaaaaaaaaaaaaaaaaaaaaaaaaa

Signed-off-by: Drapersniper <27962761+drapersniper@users.noreply.github.com>

Bakersbakebread commented on redbot/core/core_commands.py:
```diff
@@ -976,6 +979,70 @@ async def setdescription(self, ctx: commands.Context, *, description: str = ""):
+                _(
+                    "The bot will no longer store usernames and nicknames in Bank, "
+                    "ModLog and Mod cogs.\n"
+                    "To delete all stored names use `[p]set deletenames"
```
```suggestion
                    "To delete all stored names use `[p]set deletenames`"
```

Bakersbakebread commented on redbot/cogs/modlog/modlog.py:
```diff
@@ -33,6 +34,18 @@ async def reapply_audittype_migration(self, ctx: commands.Context):
+                    if "last_known_username" in modlog_data[guild_id][case_number]:
+                        del modlog_data[guild_id][case_number]["last_known_username"]
+                    await asyncio.sleep(0)
+        await ctx.tick()
```
```suggestion
        await ctx.send(_("Usernames have been deleted from the ModLog config."))
```
This `ctx.tick()` is inconsistent with:
- `bankset deletenames`
- `modset deletenames`

Drapersniper committed:
Apply suggestions from code review

Co-authored-by: Sean <29239704+Bakersbakebread@users.noreply.github.com>

Bakersbakebread reviewed (approved):
I'm not sure on other parts of the codebase, but I thought of maybe splitting out "last_known_username" into a constant somehere  ¯\\_(ツ)_/¯

Drapersniper commented:
Closing in favour of the data api PR that has been merged
